truffle/src/main/java/org/jruby/truffle/core/regexp/RegexpNodes.java

```diff
@@ -446,24 +468,83 @@ public int hash(DynamicObject regexp) {
             makeSubstringNode = RopeNodesFactory.MakeSubstringNodeGen.create(null, null, null);
         }
 
         @Specialization(guards = "isRubyString(string)")
-        public Object matchStart(DynamicObject regexp, DynamicObject string, int startPos) {
-            final Object matchResult = matchCommon(getContext(), this, makeSubstringNode, regexp, string, false, false, startPos);
+        public Object matchStart(VirtualFrame frame, DynamicObject regexp, DynamicObject string, int startPos) {
+            final DynamicObject dupedString = (DynamicObject) dupNode.call(frame, string, "dup");
+            final Object matchResult = matchCommon(getContext(), this, makeSubstringNode, regexp, dupedString, false, startPos);
+
             if (RubyGuards.isRubyMatchData(matchResult) && Layouts.MATCH_DATA.getRegion((DynamicObject) matchResult).numRegs > 0
                 && Layouts.MATCH_DATA.getRegion((DynamicObject) matchResult).beg[0] == startPos) {
                 return matchResult;
             }
+
             return nil();
         }
     }
 
+    @CoreMethod(names = "last_match", onSingleton = true, optional = 1, lowerFixnum = 1)
+    public abstract static class LastMatchNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public Object lastMatch(NotProvided index) {
+            return getMatchData();
+        }
+
+        @Specialization
+        public Object lastMatch(VirtualFrame frame, int index,
+                                @Cached("create()") MatchDataNodes.GetIndexNode getIndexNode) {
+            final Object matchData = getMatchData();
+
+            if (matchData == nil()) {
+                return nil();
+            } else {
+                return getIndexNode.executeGetIndex(frame, matchData, index, NotProvided.INSTANCE);
+            }
+        }
+
+        @TruffleBoundary
+        private Object getMatchData() {
+            Frame frame = getContext().getCallStack().getCallerFrameIgnoringSend().getFrame(FrameInstance.FrameAccess.READ_WRITE, true);
+            FrameSlot slot = frame.getFrameDescriptor().findFrameSlot("$~");
+
+            while (slot == null) {
+                final Frame nextFrame = RubyArguments.getDeclarationFrame(frame);
+
+                if (nextFrame == null) {
+                    return nil();
+                } else {
+                    slot = nextFrame.getFrameDescriptor().findFrameSlot("$~");
+                    frame = nextFrame;
+                }
+            }
+
+            final Object previousMatchData;
+            try {
+                previousMatchData = frame.getObject(slot);
+
+                if (previousMatchData instanceof ThreadLocalObject) {
+                    final ThreadLocalObject threadLocalObject = (ThreadLocalObject) previousMatchData;
+
+                    return threadLocalObject.get();
+                } else {
+                    return previousMatchData;
```

Comment:
I was trying to not hide the issue if a bad value were stored. I guess I could just add an assertion.